### PR TITLE
Partialslip and freeslip interpolation

### LINKED ISSUE #1050: Implement partial slip boundary interpolation in JIT mode
In the [Preventing stuck particles documentation](https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/documentation_unstuck_Agrid.ipynb), @reint-fischer proposed a partial slip boundary condition. This required him to write a special Kernel (cells 33 and 34 in the notebook) that currently only works on rectangular grids. 

However, implementing the interpolation itself inside `parcels.h` (for JIT) and in `field.py` for Scipy would allow this partial-slip under the hood, and therefore make it much more generally applicable and easier to use. It would be an alternative to the implementation of e.g. the `'linear_invdist_land_tracer'` interpolation method implemented in #815

### PR DESCRIPTION
**Pull request by erikvansebille**
This PR implements the `interp_method='partialslip'` in JIT and Scipy, as suggested by @reint-fischer in the [Preventing stuck particles documentation](https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/documentation_unstuck_Agrid.ipynb) and thereby fixes #1050

### CODE REVIEW

File: parcels/field.py [hunk 1]
```diff
@@ -1635,8 +1635,91 @@ def spatial_c_grid_interpolation3D(self, ti, z, y, x, time, particle=None):
             w = self.W.units.to_target(w, x, y, z)
         return (u, v, w)
 
+    def _is_land2D(self, di, yi, xi):
+        if self.U.data.ndim == 3:
+            if di < np.shape(self.U.data)[0]:
+                return (self.U.data[di, yi, xi] == 0.) and (self.V.data[di, yi, xi] == 0.)
```
Comment (CKehl) [inline parcels/field.py:1641]: "float == 0"-check error. Please replace with "np.isclose(0)".

File: parcels/field.py [hunk 2]
```diff
@@ -1635,8 +1635,91 @@ def spatial_c_grid_interpolation3D(self, ti, z, y, x, time, particle=None):
...
+            else:
+                return True
+        else:
+            if di < self.U.grid.zdim and yi < np.shape(self.U.data)[-2] and xi < np.shape(self.U.data)[-1]:
+                return (self.U.data[0, di, yi, xi] == 0.) and (self.V.data[0, di, yi, xi] == 0.)
```
Comment (CKehl) [inline parcels/field.py:1646]: "float == 0"-check error. Please replace with "np.isclose(0)".

File: parcels/include/parcels.h [hunk 3]
```diff
@@ -954,6 +954,265 @@ static inline StatusCode temporal_interpolationUVW_c_grid(type_coord x, type_coo
   }
 }
 
+static inline StatusCode calculate_slip_conditions_2D(double xsi, double eta, float dataU[2][2],
+                                                      float dataV[2][2], float dataW[2][2],
+                                                      float *u, float *v, float *w, int interp_method, int withW)
+{
+      float f_u = 1, f_v = 1, f_w = 1;
+      if ((is_zero_flt(dataU[0][0])) && (is_zero_flt(dataU[0][1])) &&
+          (is_zero_flt(dataV[0][0])) && (is_zero_flt(dataV[0][1])) && eta > 0.){
```
Comment (CKehl) [inline parcels/include/parcels.h:963]: good that you used our numerically-stable 0-check.

File: parcels/include/parcels.h [hunk 4]
```diff
@@ -954,6 +954,265 @@ static inline StatusCode temporal_interpolationUVW_c_grid(type_coord x, type_coo
...
+      *u *= f_u;
+      *v *= f_v;
+      if (withW) {
+        *w *= f_w;
+      }
+
+  return SUCCESS;
+}
+
+static inline StatusCode temporal_interpolation_slip(type_coord x, type_coord y, type_coord z, double time, CField *U, CField *V, CField *W,
```
Comment (CKehl) [inline parcels/include/parcels.h:1123]: function looks very neat and tidy - nice.

File: parcels/field.py (hunk 1 above)
Comment (CKehl) [reply parcels/field.py:1641]: function signature:
```
numpy.isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False)
```

might need to set `equal_nan=True` in this case.

**Review by CKehl — CHANGES_REQUESTED** [review]
minor change request in the field.py.
I didn't check the notebook cause I couldn't view it in the browser. I trust the oceanographers can give better feedback on it (and will do so) than I can.

File: parcels/field.py (hunk 1 above)
Comment (erikvansebille) [reply parcels/field.py]: Fixed now

File: parcels/include/parcels.h (hunk 3 above)
Comment (erikvansebille) [reply parcels/include/parcels.h:963]: Thanks!

File: parcels/field.py (hunk 2 above)
Comment (erikvansebille) [reply parcels/field.py]: Fixed now

File: parcels/include/parcels.h (hunk 4 above)
Comment (erikvansebille) [reply parcels/include/parcels.h:1123]: Thanks!

**Review by CKehl — APPROVED** [review]

**Review by reint-fischer — APPROVED** [review]
This looks really good and complete! Very nice to see the full implementation and the notebook is very easy to understand now I think. Interested to see how these impact study results!

File: parcels/field.py [hunk 5]
```diff
@@ -1635,8 +1635,91 @@ def spatial_c_grid_interpolation3D(self, ti, z, y, x, time, particle=None):
...
+        (xsi, eta, zeta, xi, yi, zi) = self.U.search_indices(x, y, z, ti, time, particle=particle)
+        di = ti if self.U.grid.zdim == 1 else zi  # general third dimension
+
+        f_u, f_v, f_w = 1, 1, 1
+        if self._is_land2D(di, yi, xi) and self._is_land2D(di, yi, xi+1) and self._is_land2D(di+1, yi, xi) \
+                and self._is_land2D(di+1, yi, xi+1) and eta > 0:
+            if self.U.interp_method == 'partialslip':
+                f_u = f_u * (.5 + .5 * eta) / eta
+                if self.vector_type == '3D':
+                    f_w = f_w * (.5 + .5 * eta) / eta
```
[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/field.py:1660]: Thanks!

File: parcels/field.py [hunk 6]
```diff
@@ -1635,8 +1635,91 @@ def spatial_c_grid_interpolation3D(self, ti, z, y, x, time, particle=None):
             w = self.W.units.to_target(w, x, y, z)
         return (u, v, w)
 
+    def _is_land2D(self, di, yi, xi):
+        if self.U.data.ndim == 3:
+            if di < np.shape(self.U.data)[0]:
```
[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/field.py:1640]: This is needed if one does 3D Advection on a field with only 1 depth layer (perhaps a bit of a construed usecase; but they exist in the unit tests). That's why this check is needed

**Pull request merged**